File: backend/app/schemas/analytics.py
from pydantic import BaseModel


class NamedValue(BaseModel):
    name: str
    value: float


class TopArtistItem(BaseModel):
    id: str
    name: str
    popularity: int
    genres: list[str] = []
    image_url: str | None = None


class TopTrackItem(BaseModel):
    id: str
    name: str
    artist: str
    popularity: int
    duration_ms: int


class AudioFeaturesAnalysis(BaseModel):
    danceability: float
    energy: float
    loudness: float
    valence: float
    acousticness: float
    instrumentalness: float
    tempo: float
    # Normalized 0–1 for radar chart
    danceability_norm: float
    energy_norm: float
    loudness_norm: float
    valence_norm: float
    acousticness_norm: float
    instrumentalness_norm: float
    tempo_norm: float


class ListeningTimeStats(BaseModel):
    total_ms: int
    total_hours: float
    track_count: int


class AnalyticsDashboard(BaseModel):
    time_range: str
    top_artists: list[TopArtistItem]
    top_tracks: list[TopTrackItem]
    genre_distribution: list[NamedValue]
    artist_popularity: list[NamedValue]
    monthly_listening: list[NamedValue]
    listening_time: ListeningTimeStats
    audio_features: AudioFeaturesAnalysis | None

File: frontend/src/types/analytics.ts
import type { TimeRange } from './index'

export interface NamedValue {
  name: string
  value: number
}

export interface TopArtistAnalytics {
  id: string
  name: string
  popularity: number
  genres: string[]
  image_url: string | null
}

export interface TopTrackAnalytics {
  id: string
  name: string
  artist: string
  popularity: number
  duration_ms: number
}

export interface AudioFeaturesAnalysis {
  danceability: number
  energy: number
  loudness: number
  valence: number
  acousticness: number
  instrumentalness: number
  tempo: number
  danceability_norm: number
  energy_norm: number
  loudness_norm: number
  valence_norm: number
  acousticness_norm: number
  instrumentalness_norm: number
  tempo_norm: number
}

export interface ListeningTimeStats {
  total_ms: number
  total_hours: number
  track_count: number
}

export interface AnalyticsDashboard {
  time_range: TimeRange
  top_artists: TopArtistAnalytics[]
  top_tracks: TopTrackAnalytics[]
  genre_distribution: NamedValue[]
  artist_popularity: NamedValue[]
  monthly_listening: NamedValue[]
  listening_time: ListeningTimeStats
  audio_features: AudioFeaturesAnalysis | null
}
